[Incomplete] Friendly Biome IDs

justinrusso opened the pull request:
For #159

This adds support for referencing a biome using its resource location string, rather than the numerical id of the biome. Using id caused issues when adding a new mod that adds biomes, as IDs would be different between an existing world and a new world (just like the issues with item ids).

I've kept support in for using an id, but when the file is ever saved, it will over-write them with the resource location.

justinrusso committed:
Allow for resource locations to be used for biome id

Still allow the use of integers as well, but if it cannot be parsed, assume its a resource location and try to obtain the ID from it.
Will need testing to see if this logic works OK.

justinrusso committed:
Convert biome ID to resource location string for saving

TehNut commented on src/main/java/hunternif/mc/atlas/client/BiomeTextureConfig.java:
```diff
@@ -41,8 +42,41 @@ protected void loadData(JsonObject json, BiomeTextureMap data, int version) {
+					biomeResourceLocation = new ResourceLocation(key);
+				}
+
+				Biome biome = Biome.REGISTRY.getObject(biomeResourceLocation);
```
```java
Biome biome = ForgeRegistries.BIOMES.getValue(new ResourceLocation(key));
```

Use the Forge provided classes rather than the Vanilla ones.

Everything else above this is unnecessary.

TehNut commented on src/main/java/hunternif/mc/atlas/client/BiomeTextureConfig.java:
```diff
@@ -76,7 +110,8 @@ protected void saveData(JsonObject json, BiomeTextureMap data) {
 			// The rest goes into ExtTileTextureConfig
 			if (biomeID >= 0 && biomeID < 256) {
-				json.addProperty(String.valueOf(biomeID), data.textureMap.get(biomeID).name);
+				String key = Biome.REGISTRY.getObjectById(biomeID).getRegistryName().toString();
```
```java
String key = ((ForgeRegistry<Biome>) ForgeRegistries.BIOMES).getValue(biomeID);
```

Use the Forge provided classes rather than Vanilla ones.

justinrusso committed:
Use the Forge provided classes rather than Vanilla ones

justinrusso commented:
Just realizing this does not fix the issue yet, but at least makes the config easier.

Since the biome texture config is loaded at preInit and then used everywhere else as the ID, they wont be the same IDs for an old world compared to a new one. Will have to either have to use the biome in more places, or make it so when a world is loaded, the IDs are "re-generated"

justinrusso committed:
Use Biome as the key for each texture instead of its ID

When getting or setting a texture the ID is used to obtain the biome itself. This *should* be all that is needed to prevent issues with worlds that have different biome IDs.

justinrusso committed:
Use biome's resource location in logging where possible

justinrusso commented:
I feel as though this has gotten too complex for my knowledge. The real fix (at least to my understanding) would be to essentially pass around the Biome, or something of that sort rather than the ID which all methods currently accepts. This means revamping the data sent as well, which is where I started to really struggle to progress further with.

I havent worked on this in a while, so I dont recall the current state, but even just allowing the json to be the resource location rather than the `int id` would be of use (which is what I've gotten done so far)

TehNut commented:
I can take a look in the next few days if @Kenkron doesn't get to it first.

TehNut commented:
Looking into it a bit more, the best move would be to switch from using the numeric ID everywhere to the registry name. That, however, would require a version bump on the `biome_textures` file, which would break all existing configs. If that's not wanted for 1.12, it should definitely be done for 1.13.

tyra314 commented:
I'm sorry this wasn't processed earlier, but with the merge of #207 the changes of this PR are merged already. We still like to thank you for your efforts.

Maybe that the feature is nonetheless now included in the mod is a consolation for you :(
